.github/workflows/codeql-analysis.yml

```diff
@@ -266,3 +266,14 @@ jobs:
             echo "::error::Rec clang-tidy failed"
             exit 1
           fi
+
+  check-for-binaries:
+    runs-on: ubuntu-22.04
+    name: Force failure in case there are binaries present in a pull request
+    if: ${{ github.event_name == 'pull_request' }}
```

Comment:
Would it make sense to enable this test for push events as well?

Reply:
I thought about it but I did not find a clean way to achieve it.

If my understanding is correct, in case of a push event, this command `git diff HEAD^..HEAD` will only compare the latest commit with the previous one (for a PR, GitHub creates a special commit including all changes). In case of pushing more than one commit at the same time, that diff no longer works as expected. This is already happening during the `analyze` step.

Reply:
I guess that would look like a very different check - scan the whole tree for binaries, check them against an allowlist. Does not seem worth the work I think

Reply:
So when we push to the branch used in a pull request we don't get a new "special" merge commit? That's awful, it explains some things I have seen in the analyze step indeed, but it also means we cannot easily get the changes introduced by a PR which seems like a very bad design choice to me. I'll look into it.

Reply:
On the PR side, you do get a new special merge commit, covering the entire PR.

Reply:
> On the PR side, you do get a new special merge commit, covering the entire PR.

Even on a push to the branch after the PR has been opened?

Reply:
yes

Reply:
My tests seem to agree, nice! So then I would expect `git diff HEAD^..HEAD` to always provide the difference between the current HEAD of target branch (master, mostly) and the HEAD of the PR branch, which would include all commits on the PR branch. Still I'm quite sure I have seen what Alexis described earlier, where on a push to a branch for which a PR already exists, `git diff HEAD^..HEAD` only return the changes of the last commit. Although, I'm now wondering if I somehow got confused and looked at the CI run on my repository and not the one for the PR on PowerDNS/pdns.

Reply:
Looking at your repo would definitely explain that!

Reply:
It wouldn't hurt, but it would not be too useful - if I push a binary to a branch, it gets detected; then I push another thing (without adding another binary), and the branch status for this check would go green. Meanwhile the PR check would stay red as it checks the entire PR.

Reply:
Ah, I think I might be confused about something else then! With the `github.event_name == 'pull_request'` condition, is this check executed if I push a commit to a branch corresponding to an existing PR?

Reply:
yes, it will!

Reply:
(event really is the wrong word there)

Reply:
> For example, if no activity types are specified, the workflow runs when a pull request is opened or reopened or when the head branch of the pull request is updated.

Riiiiiiiiiiiiiiight. Thank you, now I wonder how I managed to misunderstand that for so long :facepalm: